[lerxuann] iP

lerxuann opened the pull request:
# Chatty : the improved Duke

> "Your mind is for having ideas, not holding them." - David Allan [(source)](https://dansilvestre.com/productivity-quotes/)

Project assignments, presentations, birthdays, you've got all these on your plate, and you think you're juggling it well, until this text message:
**You're going for your cousin's wedding later, right?** 🥰 

_Uh oh!_

Chatty helps you keep track of the things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it
2. double-click it
3. add your tasks
4. let it manage your tasks for you 😉 

Features:
- [ ] Managing tasks
- [ ] Witty, sarcastic responses that will make you want to start on your tasks well before the deadline!




If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

lerxuann committed:
Level 0: Rename, Greet, Exit

lerxuann committed:
Level 1: Echo

lerxuann committed:
Level 2: Add, List

lerxuann committed:
Level 3: Mark as Done

lerxuann committed:
Level 4: ToDos, Events, Deadlines

lerxuann committed:
A-TextUiTesting

lerxuann committed:
Level 5: Handle Errors

lerxuann committed:
Level 6: Delete & Handled more exceptions

lerxuann committed:
Level 7: Save

lerxuann committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Level 7: Save

lerxuann committed:
add LocalDateTime to tasks

lerxuann committed:
Merge branch 'branch-Level-8'

lerxuann committed:
Organise into packages, add more OOP

lerxuann committed:
Merge branch 'A-MoreOOP'

lerxuann committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

lerxuann committed:
Merge branch 'branch-A-Gradle'

lerxuann committed:
add JUnit tests

lerxuann committed:
add JUnit tests

lerxuann committed:
fix error in done status of saved tasks

lerxuann committed:
package app as JAR file

lerxuann committed:
add find command to search for tasks

lerxuann committed:
add Javadoc comments to code

lerxuann committed:
tweak code to comply with coding standards

lerxuann committed:
Merge branch 'branch-Level-9'

lerxuann committed:
Merge branch 'branch-A-CodingStandard'

lerxuann committed:
Merge branch 'branch-A-Javadoc'

lipwei1808 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+    private Ui ui;
+
+    /**
+     * Constructor for the Duke bot.
+     * @param filePath Path to data file with saved tasks from user's previous sessions
```
I think there usually is a space between the params and the header of the comment right

lipwei1808 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+        tasks.remove(t);
+        String returnLine = "Looks like you have more time to sleep now. Deleted this for you:\n"
+                + t.toString()
+                + "\nYou now have " + tasks.size() + " things to do.\n";
```
Should have a new line after a "+"?

lipwei1808 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Prints a goodbye message when users close the bot.
+     */
+    public void goodbye() {
```
Should use verbs for method names? Not sure if goodbye is a verb

lipwei1808 commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,31 @@
+        storage.writeData(tasks.toWriteString());
+        System.out.println("Ok. Your tasklist has grown longer with this addition:\n"
+        + deadline.toString()
+        + "\nYou now have " + tasks.getLength() + " things to do.\n");
```
A newline after every "+"?

lipwei1808 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+     */
+    public TaskList(ArrayList<String[]> inputList) {
+        tasks = new ArrayList<>();
+        for (String[] line : inputList) {
```
You have an extra space before the : right?

lipwei1808 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,65 @@
+        try {
+            FileWriter fw = new FileWriter(this.filePath);
+            for (String t: inputList) {
+                fw.write(t.strip()+"\n");
```
Should have spaces before and after the +

Eola-Z commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,31 @@
+        this.description = description;
+        this.by = by;
+    }
+    public boolean isExit() {
+        return false;
+    }
```
Is a method here necessary if the output is always the same?

Eola-Z commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+    public TaskList(ArrayList<String[]> inputList) {
+        tasks = new ArrayList<>();
+        for (String[] line : inputList) {
+            switch (line[0]) {
+            case "T":
+                ToDo t = new ToDo(line[2]);
+                if (line[1].equals("1")) {
+                    t.markAsDone();
+                }
+                tasks.add(t);
+                break;
+            case "D":
+                Deadline d = new Deadline(line[2],
+                        LocalDateTime.parse(line[3], DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm")));
+                if (line[1].equals("1")) {
+                    d.markAsDone();
+                }
+                tasks.add(d);
+                break;
+            case "E":
+                Event e = new Event(line[2],
+                        LocalDateTime.parse(line[3], DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm")),
+                        LocalDateTime.parse(line[4], DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm")));
+                if (line[1].equals("1")) {
+                    e.markAsDone();
+                }
+                tasks.add(e);
+                break;
+            }
```
Good job following the coding standard on switch cases! 👍

Eola-Z commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+     * Returns if app should quit after executing this command.
+     * @return true if user inputs "bye", false otherwise
+     */
+    public abstract boolean isExit();
```
As stated previously, is it necessary for this to be a method?

Eola-Z commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+     * Returns a string representation of the tasks in the tasklist.
+     * @return Tasks to write into save file
+     */
+    public String[] toWriteString() {
```
I noticed that you used the same name 'toWriteString' for various methods. Some return a String, others return a String array.
Is there a way to differentiate these method names so that they won't be as confusing?

Also, 'toWriteString' doesn't quite sound like a verb. I presume it's based on 'toString', but is there a way to name these functions so they are more descriptive?

Eola-Z commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Displays a line for presentation.
+     */
+    public void showLine() {
+        System.out.println("______________________________________\n");
+    }
```
Are the horizontal lines in 'greet' necessary given that this method exists?
For instance, could I do something similar to this?
```java
public void greet() {
    showLine();
    System.out.println("....");
    showLine();
}
```

Eola-Z reviewed (commented):
Overall, I don't have much to add that hasn't already been said by the other reviewer. And I think these issues are incredibly minor nitpicks anyway.

This iP is really impressive! You have done a great job with it. 👍 👍 👍 
<sub>Better than anything I've managed, that's for sure... 🤡 </sub>

lerxuann committed:
add GUI to the chatbox with JavaFX technology

lerxuann committed:
Merge branch 'branch-Level-10'

lerxuann committed:
add assertions in Parser, Storage and TaskList

lerxuann committed:
improve code quality

lerxuann committed:
Merge pull request #1 from lerxuann/branch-A-Assertions

add assertions in Parser, Storage and TaskList

Let's add assertions to some of the classes that contain most of the logic. This helps document important assumptions that should hold at various points in the code.

For example, checking if a save file exists after creating it in the Storage class.

lerxuann committed:
Merge branch 'master' into branch-A-CodeQuality

lerxuann committed:
Merge pull request #2 from lerxuann/branch-A-CodeQuality

improve code quality

Let's clean up some code and improve on code quality. For example, in Ui.java, some methods are no longer used, and in the Command classes, the 'isEnd' method does not have to be overridden in many classes.

lerxuann committed:
add simple help function

Let's provide some in-app guidance to users, especially since first time users will not know what commands the app supports. As the list of app features grows, it would also be hard for users to remember every function. Thus, providing a help functiion would save users time, and teach them how to use the app.

lerxuann committed:
Merge branch 'branch-C-Help'

lerxuann committed:
add greeting when app is launched

Let's have the bot greet the user when the app is first launched. It will also direct new users to use the 'help' function so that they know the list of available commands.

lerxuann committed:
improve GUI

lerxuann committed:
add lines to conform to personality

lerxuann committed:
improve OOP of listing tasks from tasklist

lerxuann committed:
fix bug in saving deadline and event dates

lerxuann committed:
add screenshot of UI

lerxuann committed:
handle more errors

lerxuann committed:
Merge remote-tracking branch 'origin/master'

lerxuann committed:
fix bug in creation of save file

lerxuann committed:
handle minor input errors

lerxuann committed:
Update README.md

lerxuann committed:
fix minor bug when user inputs event
